src/algorithms/tonal/pitch2midi.cpp

```diff
@@ -0,0 +1,297 @@
+#include "pitch2midi.h"
+#include "essentiamath.h"
+
+using namespace std;
+using namespace essentia;
+using namespace standard;
+
+const char* Pitch2Midi::name = "Pitch2Midi";
+const char* Pitch2Midi::category = "Pitch";
+const char *Pitch2Midi::description = DOC("This algorithm estimates the midi note ON/OFF detection from raw pitch and voiced values, using midi buffer and uncertainty checkers.");
+
+void Pitch2Midi::configure()
+{
+  _sampleRate = parameter("sampleRate").toReal();
+  _hopSize = parameter("hopSize").toInt();
+  _minFrequency = parameter("minFrequency").toReal();
+  _minOcurrenceRate = parameter("minOcurrenceRate").toReal();
+  _bufferDuration = parameter("midiBufferDuration").toReal();
+  _minOnsetCheckPeriod = parameter("minOnsetCheckPeriod").toReal();
+  _minOffsetCheckPeriod = parameter("minOffsetCheckPeriod").toReal();
+  _minNoteChangePeriod = parameter("minNoteChangePeriod").toReal();
+  _applyCompensation = parameter("applyTimeCompensation").toBool();
+  // former Pitch2Midi only parameters
+  _tuningFreq = parameter("tuningFreq").toReal();
+  _transposition = parameter("transpositionAmount").toInt();
+
+  _frameTime = _hopSize / _sampleRate;
+  _minOnsetCheckThreshold = _minOnsetCheckPeriod / _frameTime;
+  _minOffsetCheckThreshold = _minOffsetCheckPeriod / _frameTime;
+  _minNoteChangeThreshold = _minNoteChangePeriod / _frameTime;
+
+  _unvoicedFrameCounter = 0;
+  _offsetCheckCounter = 0;
+  _onsetCheckCounter = 0;
+    
+  _minOcurrenceRatePeriod = _minOcurrenceRate * _bufferDuration;
+  _minOcurrenceRateThreshold = _minOcurrenceRatePeriod / _frameTime;
+
+  _capacity = capacity();
+  _framebuffer = AlgorithmFactory::create("FrameBuffer");
+  _framebuffer->configure("bufferSize", _capacity);
+
+}
+
+void Pitch2Midi::getMidiNoteNumber(Real pitch)
+{
+  _detectedPitch = pitch;
+    
+  if (pitch <= 0) { _detectedPitch = 1e-05; }
+  int idx = hz2midi(pitch, _tuningFreq);
+  _midiNoteNumberTransposed = idx + _transposition;
+}
+
+int Pitch2Midi::capacity()
+{
+  float hopSizeFloat = _hopSize; // ensure no int/int division happens
+  float sampleRateFloat = _sampleRate;
+  int c = static_cast<int>( round( sampleRateFloat / hopSizeFloat * _bufferDuration ) );
+  return max(_minCapacity, c);
+}
+
+// this should NOT be called until framebuffer.compute has been called
+bool Pitch2Midi::hasCoherence()
+{
+  Real sum = accumulate(_buffer.begin(), _buffer.end(), 0.0);
+  if (sum / _capacity == _buffer[0]) {
+      return true;
+  }
+  return false;
```

Comment:
We are checking if the average note is the same as the latest note in the buffer, however, the average does not guarantee note stability (same note through the entire buffer), if this is what we are looking for here.

Reply:
No, it is checking if the entire buffer has the same note, coherence. In those cases, it is clear there is a note (maybe inside, maybe the starting). Average could be used when you have open a new note from zero, that's what minOcurrenceRate does.

Reply:
I think I did a flow diagram you should check I will share it with you.